### PR TITLE
v1, working base api and dashboard

### PR DESCRIPTION
**Pull request by BilalG1**

### CODE REVIEW

File: tarsier-service/next/public/next.svg
Comment (asim-shrestha) [inline tarsier-service/next/public/next.svg:1]: Let's remove the boilerplate

File: tarsier-service/next/public/vercel.svg
Comment (asim-shrestha) [inline tarsier-service/next/public/vercel.svg:1]: Also remove

File: tarsier-service/next/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+- [Next.js Documentation](https://nextjs.org/docs) - learn about Next.js features and API.
+- [Learn Next.js](https://nextjs.org/learn) - an interactive Next.js tutorial.
+
+You can check out [the Next.js GitHub repository](https://github.com/vercel/next.js/) - your feedback and contributions are welcome!
+
+## Deploy on Vercel
+
+The easiest way to deploy your Next.js app is to use the [Vercel Platform](https://vercel.com/new?utm_medium=default-template&filter=next.js&utm_source=create-next-app&utm_campaign=create-next-app-readme) from the creators of Next.js.
+
+Check out our [Next.js deployment documentation](https://nextjs.org/docs/deployment) for more details.
```
Comment (asim-shrestha) [inline tarsier-service/next/README.md:36]: Remove boilerpllate

File: tarsier-service/next/src/app/layout.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import type { Metadata } from "next";
+import { Inter } from "next/font/google";
+import "./globals.css";
+
+const inter = Inter({ subsets: ["latin"] });
+
+export const metadata: Metadata = {
+  title: "Create Next App",
+  description: "Generated by create next app",
```
Comment (asim-shrestha) [inline tarsier-service/next/src/app/layout.tsx:9]: nit: update

File: tarsier-service/next/src/components/apiLogsTable.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import {
+  Table,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableHeader,
+  TableRow,
+} from "@/components/ui/table"
+
+const BACKEND_BASE_URL = "http://localhost:8000" //import
```
Comment (asim-shrestha) [inline tarsier-service/next/src/components/apiLogsTable.tsx:11]: should steal the env management of create-t3-app (Can look at AgentGPT for this as well)

Basically it defines both a client and server env (though we don't really need the server env) and fails startup/deployment if required values aren't set

File: tarsier-service/service/pageText.txt [hunk 4]
```diff
@@ -0,0 +1,96 @@
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+Product ✓ Solutions         Resources       Open Source ✓     Enterprise      Pricing                                                                             Q Search or jump to ...                        Sign in     Sign up
```
Comment (asim-shrestha) [inline tarsier-service/service/pageText.txt:2]: What is this for?

File: tarsier-service/service/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+python-dotenv = "^1.0.1"
+tarsier = "^0.6.38"
+openai = "^1.41.0"
+sqlalchemy = "^2.0.32"
+pymysql = "^1.1.1"
+cryptography = "^43.0.0"
+propelauth-fastapi = "^2.1.16"
+stripe = "^10.8.0"
+
+
```
Comment (asim-shrestha) [inline tarsier-service/service/pyproject.toml:25]: nit: we should add formatting and type checking like ruff and mypy (Can consult agentgpt for current settings this)

File: tarsier-service/service/src/auth.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import os
+from fastapi import Depends, HTTPException, Security
+from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer
+from propelauth_fastapi import init_auth, User
+from dotenv import load_dotenv
+
+
+load_dotenv() 
+auth = init_auth("https://6966894145.propelauthtest.com", os.environ["PROPEL_AUTH_API_KEY"])
```
Comment (asim-shrestha) [inline tarsier-service/service/src/auth.py:9]: Should place secrets in Pydantic settings for easier management. This is what we do both currently and on AgentGPT https://docs.pydantic.dev/latest/concepts/pydantic_settings/#usage

File: tarsier-service/service/src/browser.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        chromium = playwright.chromium
+        browser = await chromium.connect_over_cdp('wss://connect.browserbase.com?apiKey='+ os.environ["DEWORKD_BROWSERBASE_API_KEY"])
+        context = browser.contexts[0]
+        page = context.pages[0]
+        await page.goto(url)
+        page_text, tag_to_xpath = await tarsier.page_to_text(page, tagless=True)
+        return page_text
+
+
+# asyncio.run(run())
```
Comment (asim-shrestha) [inline tarsier-service/service/src/browser.py:30]: nit: remove

File: tarsier-service/service/src/llm.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+structured_data_sys_prompt = '''
+You are a data extraction agent. 
+You will be given a textual representation of a website and your job is to extract data according the the given schema.
+Respond only in json with the following schema:
+{schema}
+'''
+
+async def structured_data_from_page_text(page_text: str, schema: str):
+    print(structured_data_sys_prompt.format(schema=schema))
```
Comment (asim-shrestha) [inline tarsier-service/service/src/llm.py:16]: nit: delete

File: tarsier-service/service/src/models.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
+from pydantic import BaseModel
+from typing import Literal, Optional
+
+
+class JsonSchemaValue(BaseModel):
+    type: Literal['string', 'number', 'boolean', 'array', 'object']
```
Comment (asim-shrestha) [inline tarsier-service/service/src/models.py:6]: Should document what types are available to use if not already done

File: tarsier-service/service/src/server.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+async def extract_data(data: ExtractData, user: User = Depends(validate_api_key)) -> ExtractResponseData:
+    page_text = await url_to_page_text(data.url)
+    extracted_data = await llm.structured_data_from_page_text(page_text, data.outputSchema)
+    print(extracted_data)
+
+    response = { 'data': extracted_data }
+    if (data.options and data.options.return_page_text):
+        response['page_text'] = page_text
+
+    db.create_extraction_job(user["user_id"], data.url, page_text, data.outputSchema, extracted_data)
```
Comment (asim-shrestha) [inline tarsier-service/service/src/server.py:35]: nit: Fine to leave this for now but in practice, we want all dependencies to be injected to the function as opposed to relying on global `llm` or `db` objects

File: tarsier-service/service/src/llm.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+            },
+            { 
+                "role": "user",
+                "content": page_text
+            }
+        ],
+        model="gpt-4o",
+        response_format={ "type": "json_object" }
+    )
+    return chat_completion.choices[0].message.content
```
Comment (asim-shrestha) [inline tarsier-service/service/src/llm.py:31]: We should test out the structured data extraction mode

File: tarsier-service/.DS_Store
Comment (awtkns) [inline tarsier-service/.DS_Store:1]: This should be git ignored

File: tarsier-service/next/src/app/dash/page.tsx [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  useEffect(() => {
+    const headers = { 'Authorization': `Bearer ${props.accessToken}`}
+    fetch(`${BACKEND_BASE}/job-count-cur-month`, { headers })
+      .then(r => r.json())
+      .then(count => {
+        setJobsThisMonth(count)
+      })
+  }, [])
```
Comment (awtkns) [inline tarsier-service/next/src/app/dash/page.tsx:31]: ideally down the road we'd use react-query for this

File: tarsier-service/service/pyproject.toml (hunk 5 above)
Comment (awtkns) [reply tarsier-service/service/pyproject.toml:25]: +1 ruff and mypy should be dev dependancies

File: tarsier-service/service/src/browser.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+async def url_to_page_text(url: str):
+    
+    async with async_playwright() as playwright:
+        chromium = playwright.chromium
+        browser = await chromium.connect_over_cdp('wss://connect.browserbase.com?apiKey='+ os.environ["DEWORKD_BROWSERBASE_API_KEY"])
+        context = browser.contexts[0]
+        page = context.pages[0]
+        await page.goto(url)
+        page_text, tag_to_xpath = await tarsier.page_to_text(page, tagless=True)
+        return page_text
```
Comment (awtkns) [inline tarsier-service/service/src/browser.py:27]: is the browser being closed here?

File: tarsier-service/service/src/db.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    index = conn.execute(text('''
+        SELECT COUNT(*)
+        FROM information_schema.statistics
+        WHERE TABLE_SCHEMA = DATABASE()
+        AND TABLE_NAME = 'jobs' 
+        AND INDEX_NAME = 'idx_userid_createdat'; 
+    '''))
+    if int(index.first()[0]) == 0:
+        print('creating index: ')
+        conn.execute(text('CREATE INDEX idx_userid_createdat ON jobs(user_id, createdAt)'))
```
Comment (awtkns) [inline tarsier-service/service/src/db.py:34]: This made me LOL. is annoying that there is no create index if not exists in mysql

File: tarsier-service/service/src/db.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+import json
+
+host = os.environ.get("DB_HOST") or "localhost"
+engine = create_engine(f"mysql+pymysql://user:password@{host}:3306/db") #docker compose ...@db:3306/db"
+
+with engine.connect() as conn:
+    # conn.execute(text('DROP TABLE jobs'))
+    conn.execute(text('''
+        CREATE TABLE IF NOT EXISTS jobs (
+            id VARCHAR(255) PRIMARY KEY,
```
Comment (awtkns) [inline tarsier-service/service/src/db.py:14]: nit such a large column for an pk and lead to performance issues. If you want to use UUID as a pk binary(16) is probably better. eg:

```
create table tbl_test (
    guid binary(16) default (uuid_to_bin(uuid())) not null primary key,
    name varchar(50) not null
);
```

Ideally you would also want the db to control PK generation as done in the example above. For performance reason, is also quite nice to choose a pk that is ordered by time.

I'd recommend reading the below if curious :) 
https://planetscale.com/blog/why-we-chose-nanoids-for-planetscales-api
https://planetscale.com/blog/the-problem-with-using-a-uuid-primary-key-in-mysql

File: tarsier-service/service/src/db.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+            { "id": user_id }
+        )
+        return [ row._asdict() for row in result.all() ]
+    
+def get_job_count_by_user_cur_month(user_id: str):
+    with engine.connect() as conn:
+        result = conn.execute(
+            text('''
+                SELECT COUNT(*) FROM jobs WHERE user_id = :id \
+                AND createdAt >= UNIX_TIMESTAMP(LAST_DAY(CURDATE()) + INTERVAL 1 DAY - INTERVAL 1 MONTH)
```
Comment (awtkns) [inline tarsier-service/service/src/db.py:75]: bravo on the index choice. very smart

File: tarsier-service/service/src/server.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+@app.post("/extract")
+async def extract_data(data: ExtractData, user: User = Depends(validate_api_key)) -> ExtractResponseData:
+    metadata = user["user"].get("metadata", {})
+    if not check_user_api_limit(user["user_id"], metadata):
+        raise HTTPException(status_code=401, detail="api request limit for month reached")
+
+    page_text = await url_to_page_text(data.url)
+    extracted_data = await llm.structured_data_from_page_text(page_text, data.outputSchema)
+    print(extracted_data)
+
```
Comment (awtkns) [inline tarsier-service/service/src/server.py:51]:
```suggestion
```

File: tarsier-service/service/src/db.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
+import os
+from typing import Any
+from sqlalchemy import create_engine, text, Row
+import uuid
+import json
+
+host = os.environ.get("DB_HOST") or "localhost"
+engine = create_engine(f"mysql+pymysql://user:password@{host}:3306/db") #docker compose ...@db:3306/db"
```
Comment (awtkns) [inline tarsier-service/service/src/db.py:8]: i dont think think we are doing engine.dispose on shutdown here. eg (from our prod app)

    """
    @app.on_event("shutdown")
    async def _shutdown() -> None:
        await app.state.db_primary_engine.dispose()
        await app.state.db_replica_engine.dispose()
        await shutdown_kafka(app)
"""

**Review by awtkns — COMMENTED** [review]
Looks good to me for the most part! Nice work here. My other piece of feedback would be the /extract docs could be more complete. its a little hard for others to figure out how to correctly format the schema

File: tarsier-service/service/src/db.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+with engine.connect() as conn:
+    # conn.execute(text('DROP TABLE jobs'))
+    conn.execute(text('''
+        CREATE TABLE IF NOT EXISTS jobs (
+            id VARCHAR(255) PRIMARY KEY,
+            user_id VARCHAR(60),
+            url VARCHAR(2083),
+            page_text TEXT,
+            outputSchema JSON,
```
Comment (asim-shrestha) [inline tarsier-service/service/src/db.py:18]: nit: inconsistent casing

File: tarsier-service/service/src/server.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+@app.post("/create-checkout-session")
+async def create_session(user: User = Depends(auth.require_user)):
+    try:
+        session = stripe.checkout.Session.create(
+            ui_mode = 'embedded',
+            line_items=[
+                {
+                    'price': 'price_1Pq3FAHyVxfGsTbDa9MvKiut', #recurring $10 / month
```
Comment (asim-shrestha) [inline tarsier-service/service/src/server.py:78]: nit better variable naming can get rid of the comment
```suggestion
recurring_10_a_month_price = 'price_1Pq3FAHyVxfGsTbDa9MvKiut'
...
'price': 'price_1Pq3FAHyVxfGsTbDa9MvKiut',
```

File: tarsier-service/service/Dockerfile [hunk 21]
```diff
@@ -0,0 +1,23 @@
+# Use the official Python 3.12 image as the base
+FROM python:3.12-slim
```
Comment (asim-shrestha) [inline tarsier-service/service/Dockerfile:2]: nit: not sure if 3.12 is more stable now but it's had some issues in the past. 3.11 might be the safer choice

File: tarsier-service/next/src/app/dash/page.tsx (hunk 12 above)
Comment (asim-shrestha) [reply tarsier-service/next/src/app/dash/page.tsx:31]: https://tanstack.com/query/latest/docs/framework/react/overview

File: tarsier-service/next/src/components/checkoutForm.tsx [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+    return fetch(`${BACKEND_BASE}/create-checkout-session`, {
+      headers,
+      method: "POST",
+    })
+      .then((res) => res.json())
+      .then((data) => data.clientSecret);
+  }, []);
+
+  const options = { fetchClientSecret };
+//absolute inset-0 bg-white z-50
```
Comment (asim-shrestha) [inline tarsier-service/next/src/components/checkoutForm.tsx:33]:
```suggestion
```

File: tarsier-service/service/src/llm.py [hunk 23]
```diff
@@ -0,0 +1,31 @@
+from openai import OpenAI
+from dotenv import load_dotenv
+
+
+load_dotenv()
+client = OpenAI()
```
Comment (asim-shrestha) [inline tarsier-service/service/src/llm.py:6]: not sure if this is still a problem but making multiple requests (especially concurrent requests) on a single open ai client used to be problematic. Should instantiate in the function it's being used in

**Pull request closed without merging**